src/thunderbots/software/geom/util.cpp

```diff
@@ -752,6 +756,39 @@ std::optional<Point> lineIntersection(const Vector &a, const Vector &b, const Ve
     return std::make_optional(intersection);
 }
 
+std::pair<std::optional<Point>, std::optional<Point>> raySegmentIntersection(
+    Ray &ray, Segment &segment)
+{
+    Point ray2 = ray.getRayStart() + ray.getDirection();
+
+    std::optional<Point> intersection = lineIntersection(
+        ray.getRayStart(), ray2, segment.getSegStart(), segment.getEnd());
+
+    // If there exists a single intersection, and it exists on the ray and within the
+    // segment
+    if (intersection.has_value() && contains(ray, intersection.value()) &&
+        contains(segment, intersection.value()))
+    {
+        return std::make_pair(intersection, std::nullopt);
+    }
+    // The ray and segment are parallel, and collinear
+    else if (!intersection.has_value() &&
+             collinear(ray.getRayStart(), segment.getSegStart(), segment.getEnd()))
+    {
+        // Since we know the ray and segment are overlapping, return the ray start
+        // position, and the end of the segment that is in the direction of the ray
+        ray.getDirection().norm() == (segment.getEnd() - ray.getRayStart()).norm()
```

Comment:
I might be wrong here, but this seems like it could have some issues with double precision such that this never returns true? Could we add a few more tests to check for the different overlapping cases to make sure this isn't an issue?